internal/datastore/postgres/postgres_test.go

```diff
@@ -738,6 +757,239 @@ func ConcurrentRevisionHeadTest(t *testing.T, ds datastore.Datastore) {
+
+	// Write another revision.
+	afterRev, err := ds.ReadWriteTx(ctx, func(rwt datastore.ReadWriteTransaction) error {
+		rtu := tuple.Touch(&core.RelationTuple{
+			ResourceAndRelation: &core.ObjectAndRelation{
+				Namespace: "resource",
+				ObjectId:  "2345",
+				Relation:  "reader",
+			},
+			Subject: &core.ObjectAndRelation{
+				Namespace: "user",
+				ObjectId:  "456",
+				Relation:  "...",
+			},
+		})
+		return rwt.WriteRelationships(ctx, []*core.RelationTupleUpdate{rtu})
+	})
+	require.NoError(err)
+	require.True(afterRev.GreaterThan(commitFirstRev))
+	require.True(afterRev.GreaterThan(commitLastRev))
+
+	// Ensure that the last revision is eventually seen from the watch.
+	require.Eventually(func() bool {
+		seenWatchRevisionsLock.Lock()
+		defer seenWatchRevisionsLock.Unlock()
+		return len(seenWatchRevisions) == 3 && seenWatchRevisions[len(seenWatchRevisions)-1].String() == afterRev.String()
+	}, 2*time.Second, 5*time.Millisecond)
+}
+
+func OverlappingRevisionWatchTest(t *testing.T, ds datastore.Datastore) {
+	require := require.New(t)
+
+	ctx := context.Background()
+	r, err := ds.ReadyState(ctx)
+	require.NoError(err)
+	require.True(r.IsReady)
+
+	rev, err := ds.HeadRevision(ctx)
+	require.NoError(err)
+
+	pds := ds.(*pgDatastore)
+	require.True(pds.watchEnabled)
+
+	prev := rev.(postgresRevision)
+	nextx := prev.snapshot.xmax + 1
+
+	// Manually construct an equivalent of overlapping transactions in the database, from the repro
+	// information (See: https://github.com/authzed/spicedb/issues/1272)
+	err = pgx.BeginTxFunc(ctx, pds.writePool, pgx.TxOptions{IsoLevel: pgx.Serializable}, func(tx pgx.Tx) error {
+		_, err := tx.Exec(ctx, fmt.Sprintf(
+			`INSERT INTO %s ("%s", "%s") VALUES ('%d', '%d:%d:')`,
+			tableTransaction,
+			colXID,
+			colSnapshot,
+			nextx,
+			nextx,
+			nextx,
+		))
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.Exec(ctx, fmt.Sprintf(
+			`INSERT INTO %s ("%s", "%s", "%s", "%s", "%s", "%s", "%s", "%s", "%s") VALUES ('somenamespace', '123', 'viewer', 'user', '456', '...', '', null, '%d'::xid8)`,
+			tableTuple,
+			colNamespace,
+			colObjectID,
+			colRelation,
+			colUsersetNamespace,
+			colUsersetObjectID,
+			colUsersetRelation,
+			colCaveatContextName,
+			colCaveatContext,
+			colCreatedXid,
+			nextx,
+		))
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.Exec(ctx, fmt.Sprintf(
+			`INSERT INTO %s ("xid", "snapshot") VALUES ('%d', '%d:%d:')`,
+			tableTransaction,
+			nextx+1,
+			nextx,
+			nextx,
+		))
+
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.Exec(ctx, fmt.Sprintf(
+			`INSERT INTO %s ("%s", "%s", "%s", "%s", "%s", "%s", "%s", "%s", "%s") VALUES ('somenamespace', '456', 'viewer', 'user', '456', '...', '', null, '%d'::xid8)`,
+			tableTuple,
+			colNamespace,
+			colObjectID,
+			colRelation,
+			colUsersetNamespace,
+			colUsersetObjectID,
+			colUsersetRelation,
+			colCaveatContextName,
+			colCaveatContext,
+			colCreatedXid,
+			nextx+1,
+		))
```

Comment:
Since the fix works by bumping the xid in the query forward, wouldn't it make sense to include a third write here that adds a _non_overlapping row at `nextx+2` to ensure it doesn't get skipped (It should be fine because the query uses `>=`, but just to be defensive)

Reply:
I wanted the test to exactly reproduce what the user reported; I could add another test if you like?